vunit/vivado/tcl/extract_compile_order.tcl

```diff
@@ -4,12 +4,37 @@ set output_file [lindex $argv 1]
 open_project ${project_file}
 
 set file_out [open ${output_file} w]
+
+# Generate block designs, and any IPs contained by the block designs
+generate_target {simulation synthesis} [get_files *.bd]
+
+# Open all block designs otherwise they are not visible through get_bd_designs
+foreach bd [get_files *.bd] {
+    open_bd_design ${bd}
+}
+
+# Generate all IPs not contained in a block design
 foreach ip [get_ips -filter {SCOPE == ""}] {
     generate_target {simulation synthesis} ${ip}
+}
+
+# Build compilation order for all ips
+foreach ip [get_ips] {
     foreach src_file [get_files -used_in simulation -compile_order sources -of_objects ${ip}] {
         set library [get_property LIBRARY ${src_file}]
         set file_type [get_property FILE_TYPE ${src_file}]
         puts ${file_out} "${library},${file_type},${src_file}"
     }
 }
+
+# Build compilation order for all block designs
+# -quiet is required otherwise get_bd_designs will error if there are no block designs
+foreach bd [get_bd_designs -quiet] {
+    foreach src_file [get_files -used_in simulation -compile_order sources *${bd}*] {
```

Comment:
Is `*${bd}*` necessary (and in which case why) or could it be just `${bd}`?

Reply:
Problem with adding Block Design files to simulation is that they are not a valid parameter to `-of_objects`:
```tcl
ERROR: [Common 17-679] Invalid object type, 'diagram', used with '-of_objects' switch. Supported types are: 'file fileset ip reconfig_module'.
```
thus instead, I am filtering all the existing sources in project by the name of the block design.
This requires the regex `*` any filter around the block design name as it's a match type of behavior.
```tcl
get_files -used_in simulation -compile_order sources ${bd}
WARNING: [Vivado 12-818] No files matched 'my_block_design'
```
```tcl
get_files -used_in simulation -compile_order sources *${bd}*
<PATH>/bd/my_block_design/ipshared/b752/hdl/generic_baseblocks_v2_1_vl_rfs.v
<PATH>/bd/my_block_design/ipshared/ec67/hdl/axi_infrastructure_v1_1_0.vh
<PATH>/bd/my_block_design/ipshared/ec67/hdl/axi_infrastructure_v1_1_vl_rfs.v
<PATH>/bd/my_block_design/ipshared/87d1/hdl/axi_register_slice_v2_1_vl_rfs.v
<PATH>/bd/my_block_design/ipshared/c97d/simulation/fifo_generator_vlog_beh.v
<PATH>/bd/my_block_design/ipshared/c97d/hdl/fifo_generator_v13_2_rfs.vhd
<PATH>/bd/my_block_design/ipshared/c97d/hdl/fifo_generator_v13_2_rfs.v
<PATH>/bd/my_block_design/ipshared/fab7/hdl/axi_data_fifo_v2_1_vl_rfs.v
<PATH>/bd/my_block_design/ipshared/f8f3/hdl/axi_crossbar_v2_1_vl_rfs.v
<PATH>/bd/my_block_design/ip/my_block_design_xbar_0/sim/my_block_design_xbar_0.v
<PATH>/bd/my_block_design/sim/my_block_design.vhd
<PATH>/bd/my_block_design/ip/my_block_design_s00_regslice_0/sim/my_block_design_s00_regslice_0.v
<PATH>/bd/my_block_design/ip/my_block_design_s00_data_fifo_0/sim/my_block_design_s00_data_fifo_0.v
<PATH>/bd/my_block_design/ipshared/8c02/hdl/axi_protocol_converter_v2_1_vl_rfs.v
<PATH>/bd/my_block_design/ip/my_block_design_auto_pc_0/sim/my_block_design_auto_pc_0.v
<PATH>/bd/my_block_design/ip/my_block_design_m00_regslice_0/sim/my_block_design_m00_regslice_0.v
<PATH>/bd/my_block_design/ip/my_block_design_m01_regslice_0/sim/my_block_design_m01_regslice_0.v

```

Reply:
I understand, thanks!